src/backend/distributed/planner/distributed_planner.c

```diff
@@ -616,6 +662,94 @@ CreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan, Query *origi
 }
 
 
+/*
+ * InlineCtesAndCreateDistributedPlannedStmt gets all the parameters required
+ * for creating a distributed planned statement. The function is primarily a
+ * wrapper on top of CreateDistributedPlannedStmt(), by first inlining the
+ * CTEs and calling CreateDistributedPlannedStmt() in PG_TRY() block. The
+ * function returns NULL if the planning fails on the query where eligable
+ * CTEs are inlined.
+ */
+static PlannedStmt *
+InlineCtesAndCreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan,
+										  Query *originalQuery,
+										  Query *query, ParamListInfo boundParams,
+										  PlannerRestrictionContext *
+										  plannerRestrictionContext)
+{
+	/*
+	 * We'll inline the CTEs and try distributed planning, preserve the original
+	 * query in case the planning fails and we fallback to recursive planning of
+	 * CTEs.
+	 */
+	Query *copyOfOriginalQuery = copyObject(originalQuery);
+
+	RecursivelyInlineCtesInQueryTree(copyOfOriginalQuery);
+
+	/* after inlining, we shouldn't have any inliable CTEs */
+	Assert(!QueryTreeContainsInlinableCTE(copyOfOriginalQuery));
+
+	/* simply recurse into CreateDistributedPlannedStmt() in a PG_TRY() block */
+	PlannedStmt *result = TryCreateDistributedPlannedStmt(planId, localPlan,
+														  copyOfOriginalQuery,
+														  query, boundParams,
+														  plannerRestrictionContext);
+
+	return result;
+}
+
+
+/*
+ * TryCreateDistributedPlannedStmt is a wrapper around CreateDistributedPlannedStmt, simply
+ * calling it in PG_TRY()/PG_CATCH() block. The function returns a PlannedStmt if the input
+ * query can be planned by Citus. If not, the function returns NULL and generates a DEBUG4
+ * message with the reason for the failure.
+ */
+static PlannedStmt *
+TryCreateDistributedPlannedStmt(uint64 planId, PlannedStmt *localPlan,
+								Query *originalQuery,
+								Query *query, ParamListInfo boundParams,
+								PlannerRestrictionContext *plannerRestrictionContext)
+{
+	MemoryContext savedContext = CurrentMemoryContext;
+	PlannedStmt *result = NULL;
+
+	PG_TRY();
+	{
+		result = CreateDistributedPlannedStmt(planId, localPlan, originalQuery,
+											  query, boundParams,
+											  plannerRestrictionContext);
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(savedContext);
+		ErrorData *edata = CopyErrorData();
+		FlushErrorState();
+
+		/* don't try to intercept PANIC or FATAL, let those breeze past us */
+		if (edata->elevel != ERROR)
```

Comment:
We should probably add error codes to our planner errors such that we could check them here.

Reply:
Do you suggest that as a future improvement or a  required change for this PR?

As far as I can see, the cost of not doing it is to go through distributed planning twice for queries that Citus doesn't support. Are there any additional benefits?

Reply:
We don't want to continue in case of e.g. out of memory.